dlib/dnn/loss.h

```diff
@@ -451,8 +452,35 @@ namespace dlib
 
             DLIB_CASSERT(detector_windows.size() != 0, "You can't call mmod_options's constructor with a set of boxes that is empty (or only contains ignored boxes).");
 
+            set_overlap_nms(boxes);
+        }
+
+        mmod_options(
+            const std::vector<std::vector<mmod_rect>>& boxes,
+            const double min_detector_window_overlap_iou = 0.75
```

Comment:
This kind of overloading bothers me.  I'm certain there will be confused users who call this constructor when they don't mean to.  So it would be better if the constructor had to be invoked like this:

```
mmod_options options(NO_IMAGE_PYRAMID, boxes);
```

This way it's really obvious what's happening.  You could use one of C++11's new strongly typed enum to define the NO_IMAGE_PYRAMID constant to allow safe overloading of the constructors.

Reply:
Agreed, and fixed.

Reply:
Cool, thanks.

But won't people need to type.  `mmod_options options(mmod_options::assumed_input_layer_type_t::NO_IMAGE_PYRAMID, boxes);`?  We aren't writing java here man :)

Can you make it like `mmod_options options(use_image_pyramid::yes, boxes);` ?

Reply:
Oh, ok. I thought you wanted an `enum` so that more types could be added in the future. But now I understand that the point was readability – I completely agree that `mmod_options options(true, boxes);` wouldn't be clear enough.

Fixed now.

Reply:
In fact, when you pointed out the overloading confusion issue, the first solution I was thinking of was the [named constructor idiom](https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Named_Constructor). Which, I think, would fix the problem also. So when you mentioned enum, I started to think of something else (=possibility to add more modes in the future).

Anyway, I'm perfectly fine with the present approach. Just wanted to write this for completeness.

Reply:
Right, that's what I was thinking about too.  C++11 enums seem like a better approach here though.